src/event-stream/tokens-contract-handler.ts

```diff
@@ -626,27 +649,90 @@ class TokensContractHandler {
   /**
    * Make readonly contract call
    */
-  private async makeReadOnlyContractCall(
+  async makeReadOnlyContractCall(
+    functionName: string,
+    functionArgs: ClarityValue[]
+  ): Promise<ReadOnlyContractCallResponse> {
+    const url = this.stacksNetwork.getReadOnlyFunctionCallApiUrl(
+      this.contractAddress,
+      this.contractName,
+      functionName
+    );
+    const args = functionArgs.map(arg => cvToHex(arg));
+
+    const body = JSON.stringify({
+      sender: this.address,
+      arguments: args,
+    });
+
+    let response: Response;
+
+    try {
+      response = await fetchPrivate(url, {
+        method: 'POST',
+        body,
+        headers: {
+          'Content-Type': 'application/json',
+        },
+      });
+    } catch (error: any) {
+      // In case there is a network error during fetch.
+      console.log(`Error calling read only contract function value: ${error.message}`);
+      return { found: false, retriable: true };
+    }
+
+    if (!response.ok) {
+      if (response.status >= 500) {
+        return { found: false, retriable: true };
+      }
+      const msg = await response.text().catch(() => '');
+      return {
+        found: false,
+        retriable: false,
+        error: `Error calling read-only function. Response ${response.status}: ${response.statusText}. Attempted to fetch ${url} and failed with the message: "${msg}"`,
+      };
+    }
+    try {
+      const value: ReadOnlyFunctionResponse = await response.json();
+      const parseValue = parseReadOnlyResponse(value);
```

Comment:
Is `parseReadOnlyResponse` guaranteed to only ever throw if the error is actually retriable and will eventually succeed?

Reply:
For example, I think the stacks-node has a configurable max cost for these readonly function calls, so what happens if someone writes a contract with a function that is too expensive, will it cause this to throw and retry forever?

Reply:
It will throw error in 2 situations:
1- If the response is not ok. Meaning, there was a server side issue and not an issue with the function. This check is given [here](https://github.com/hirosystems/stacks.js/blob/master/packages/transactions/src/utils.ts#L256). 
2- If the `ClarityValue` returned from the function is not formatted correctly. In that case, the errors are thrown by `deserializeCV`. For instance, [Unrecognize Clarity Type](https://github.com/hirosystems/stacks.js/blob/master/packages/transactions/src/clarity/deserialize.ts#L39), [Clarity Type unavailable](https://github.com/hirosystems/stacks.js/blob/master/packages/transactions/src/clarity/deserialize.ts#L111-L113), [Invalid Clarity name](https://github.com/hirosystems/stacks.js/blob/master/packages/transactions/src/clarity/types/tupleCV.ts#L15), [Invalid byte length for Contract Name](https://github.com/hirosystems/stacks.js/blob/master/packages/transactions/src/clarity/types/principalCV.ts#L59), [Invalid Buffer length](https://github.com/hirosystems/stacks.js/blob/master/packages/transactions/src/clarity/types/bufferCV.ts#L11), [Range Errors for uintCV](https://github.com/hirosystems/stacks.js/blob/master/packages/transactions/src/clarity/types/intCV.ts#L36-L42), and [Range Errors for intCV](https://github.com/hirosystems/stacks.js/blob/master/packages/transactions/src/clarity/types/intCV.ts#L17-L24). These errors could be an issue that will keep retrying and never resolve, considering the function will always return the same result for the same `functionArgs` provided.

It should retry in the first case. But since we’re already catering to that issue with `retriable: true` [here](https://github.com/hirosystems/stacks-blockchain-api/pull/1131/files#diff-94b60ecbb28bcac15bf49d0f6601de1d8273ff795f81ac60f19a70eed6b429a7R684-R687), so we don’t actually need to worry about this error here. 
However, for the second case, I believe I should split these lines in `try block` such that,
`const value: ReadOnlyFunctionResponse = await response.json();` would be a retriable issue, considering if the headers were sent but connection broke before sending the body to the client.
`const parseValue = parseReadOnlyResponse(value);` should not be a retriable issue. Instead, it should log the actual error found, parsing the response.
Let me know if I’m wrong here. 
Thanks!

Reply:
From what I can tell, this endpoint can return an object:
```ts
interface ReadOnlyFunctionErrorResponse {
  okay: false;
  cause: string;
}
```
What areas handle the retrying or _not_ retrying for when this is `okay: false`.

What happens if the query cost exceeds the limits of the read-only-call parameters defined at
https://github.com/stacks-network/stacks-blockchain/blob/26bfd5fcdc1f25106288f18ced05290b569f6abb/testnet/stacks-node/src/config.rs#L1477-L1481
?